### PR TITLE
Add DoesNotExist label operator.

### PR DESCRIPTION
**Pull request by bgrant0607**
@mikedanese @davidopp PTAL. This is just to demonstrate how DoesNotExist could be implemented as discussed in #15350.

However, I found when implementing it that it should already be possible to express nonexistence, using notin with an empty value set.

If we want to go forward with this, more tests are needed.

### CODE REVIEW

**Comment by davidopp** [comment]
I generally favor parsimony so I am tempted to say we should go with "notin \<empty\>" but I think what you did here will be easier for users, so I think we should go with the approach in this PR.

**Comment by bgrant0607** [comment]
I have yet to confirm whether `k notin ()` translates to an empty list or a list containing an empty string.

**Comment by bgrant0607** [comment]
Also note that it appears that multiple requirements per key are currently possible.

On the plus side, that means one could specify `k,k notin (v)` to ensure that `k` exists. 

On the minus side, I haven't had time to think about the full implications of that.

For better or worse, the documentation leaves lots of wiggle room.

**Comment by davidopp** [comment]
I think multiple requirements per key would be confusing and I'd be in favor of prohibiting it in validation. When I looked at how some of the other open-source systems handle this stuff, they all seemed to restrict to a single constraint per key.

**Comment by bgrant0607** [comment]
notin doesn't accept an empty values list:
```
func TestRequirementConstructor(t *testing.T) {
        requirementConstructorTests := []struct {
                Key     string
                Op      Operator
                Vals    sets.String
                Success bool
        }{
...
                {"x", NotInOperator, sets.NewString(), false},
```

`()` parses to a list containing the empty string:
```
               {&LabelSelector{
                        getRequirement("x", NotInOperator, sets.NewString("abc"), t),
                        getRequirement("y", InOperator, sets.NewString("jkl", "mno"), t),
                        getRequirement("z", NotInOperator, sets.NewString(""), t)},
                        "x notin (abc),y in (jkl,mno),z notin ()", true},
```

**Comment by bgrant0607** [comment]
Added more tests. Will update documentation next.

**Comment by bgrant0607** [comment]
Updated labels.md. PTAL.

File: docs/user-guide/labels.md [hunk 1]
```diff
@@ -94,23 +112,25 @@ tier != frontend
 ```
 
 The former selects all resources with key equal to `environment` and value equal to `production`.
-The latter selects all resources with key equal to `tier` and value distinct from `frontend`.
+The latter selects all resources with key equal to `tier` and value distinct from `frontend` and all resources with no labels with the `tier` key.
```
Comment (davidopp) [inline docs/user-guide/labels.md]: add a comma before "and" even though it is grammatically incorrect, to make it easier to parse this sentence

File: docs/user-guide/labels.md [hunk 2]
```diff
@@ -94,23 +112,25 @@ tier != frontend
...
 ```
 environment in (production, qa)
 tier notin (frontend, backend)
 partition
+!partition
 ```
 
 The first example selects all resources with key equal to `environment` and value equal to `production` or `qa`.
-The second example selects all resources with key equal to `tier` and values other than `frontend` and `backend`.
+The second example selects all resources with key equal to `tier` and values other than `frontend` and `backend` and all resources with no labels with the `tier` key.
```
Comment (davidopp) [inline docs/user-guide/labels.md]: same thing, comma before "and all"

File: docs/user-guide/labels.md [hunk 3]
```diff
@@ -126,38 +148,39 @@ LIST and WATCH operations may specify label selectors to filter the sets of obje
...
+#### Service and ReplicationController
 
-The set of pods that a `service` targets is defined with a label selector. Similarly, the population of pods that a `replicationcontroller` is monitoring is also defined with a label selector. For management convenience and consistency, `services` and `replicationcontrollers` may themselves have labels and would generally carry the labels their corresponding pods have in common. Labels selectors for both objects are defined in `json` or `yaml` files using `map` and only _equality-based_ requirement selectors are supported:
+The set of pods that a `service` targets is defined with a label selector. Similarly, the population of pods that a `replicationcontroller` should monitor is also defined with a label selector. For management convenience and consistency, `services` and `replicationcontrollers` may themselves have labels and would generally carry the labels their corresponding pods have in common. Labels selectors for both objects are defined in `json` or `yaml` files using maps, and only _equality-based_ requirement selectors are supported:
```
Comment (davidopp) [inline docs/user-guide/labels.md]: maybe s/monitor/manage/

"would generally carry" is pretty ambiguous; you could make it clearer that this is describing a "best practice" and not something the system does automatically

File: docs/user-guide/labels.md [hunk 4]
```diff
@@ -165,26 +188,25 @@ or
...
+```yaml
+selector:
+  matchLabels:
+    component: redis
+  matchExpressions:
+    - {key: tier, operator: In, values: [cache]}
+    - {key: environment, operator: NotIn, values: [dev]}
+```
 
+`matchLabels` is a map of `{key,value}` pairs. A single `{key,value}` in the `matchLabels` map is equivalent to an element of `matchExpressions`, whose `key` field is "key", the `operator` is "In", and the `values` array contains only "value". The requirements are ANDed. `matchExpressions` is a list of pod selector requirements. The requirements are ANDed. Valid operators include In, NotIn, Exists, and DoesNotExist. The values set must be non-empty.
```
Comment (davidopp) [inline docs/user-guide/labels.md]: Say it's a map not a multi-map, assuming we enforce that invariant in validation (which I think we should).

Comment (mikedanese) [inline docs/user-guide/labels.md]: We don't currently enforce it.

**Comment by HaiyangDING** [comment]
CC @HaiyangDING

File: pkg/labels/selector.go [hunk 5]
```diff
@@ -106,6 +107,7 @@ func NewRequirement(key string, op Operator, vals sets.String) (*Requirement, er
 			return nil, fmt.Errorf("exact match compatibility requires one single value")
 		}
 	case ExistsOperator:
+	case DoesNotExistOperator:
```
Comment (mikedanese) [inline pkg/labels/selector.go]: Why don't DoesNotExistOperator, ExistsOperator cases assert that len(vals) == 0

File: pkg/labels/selector.go [hunk 6]
```diff
@@ -513,7 +522,12 @@ func (p *Parser) parseRequirement() (*Requirement, error) {
 // in case of no operator 'in, notin, ==, =, !=' are found
```
Comment (mikedanese) [inline pkg/labels/selector.go]: comment should include !

File: pkg/labels/selector.go [hunk 7]
```diff
@@ -533,6 +548,7 @@ func (p *Parser) parseKeyAndInferOperator() (string, Operator, error) {
 func (p *Parser) parseOperator() (op Operator, err error) {
 	tok, lit := p.consume(KeyAndOperator)
 	switch tok {
+	// DoesNotExistToken shouldn't be here because it's a unary operator, not a binary operator
```
Comment (mikedanese) [inline pkg/labels/selector.go]: And ExistsToken?

File: pkg/labels/selector.go [hunk 8]
```diff
@@ -457,7 +466,7 @@ func (p *Parser) parse() ([]Requirement, error) {
 	for {
 		tok, lit := p.lookahead(Values)
 		switch tok {
-		case IdentifierToken:
+		case IdentifierToken, DoesNotExistToken:
```
Comment (mikedanese) [inline pkg/labels/selector.go]: Error message on 490 needs update. Can't comment there.

**Comment by mikedanese** [comment]
Code looks correct. Thanks for the PR!

File: pkg/labels/selector.go (hunk 7 above)
Comment (bgrant0607) [inline pkg/labels/selector.go]: There is no ExistsToken.

**Comment by bgrant0607** [comment]
Addressed feedback, except for the map vs. multi-map issue.

Restricting to a single requirement per key would be a breaking API change, unless the change were only made for Job and not for LIST/WATCH filters.

My other concern about that restriction is that it would preclude some useful combinations. We could do smarter validation, such as preventing simultaneous specification of In and NotIn for the same key. I'd prefer that be done in a separate PR, though, since I have other fires burning.

File: pkg/labels/selector.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/labels/selector.go]: Oops. Missed this one. Will do.

**Comment by mikedanese** [comment]
LGTM, assigning to @davidopp

File: pkg/labels/selector.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/labels/selector.go]: Done.

File: pkg/labels/selector.go (hunk 8 above)
Comment (bgrant0607) [inline pkg/labels/selector.go]: Done.

**Comment by bgrant0607** [comment]
I'll squash after final lgtm.

**Comment by bgrant0607** [comment]
All green.

File: docs/user-guide/labels.md [hunk 9]
```diff
@@ -94,23 +112,25 @@ tier != frontend
...
 The former selects all resources with key equal to `environment` and value equal to `production`.
-The latter selects all resources with key equal to `tier` and value distinct from `frontend`.
+The latter selects all resources with key equal to `tier` and value distinct from `frontend`, and all resources with no labels with the `tier` key.
 One could filter for resources in `production` excluding `frontend` using the comma operator: `environment=production,tier!=frontend`
 
 
 ### _Set-based_ requirement
 
-_Set-based_ label requirements allow filtering keys according to a set of values. Matching objects must have all of the specified labels (i.e. all keys and at least one of the values specified for each key). Three kinds of operators are supported: `in`,`notin` and exists (only the key identifier). For example:
+_Set-based_ label requirements allow filtering keys according to a set of values. Three kinds of operators are supported: `in`,`notin` and exists (only the key identifier). For example:
```
Comment (davidopp) [inline docs/user-guide/labels.md]: exists should be in backticks

File: docs/user-guide/labels.md (hunk 4 above)
Comment (davidopp) [inline docs/user-guide/labels.md]: Maybe it's redundant but should you say that the result of matchLabels and matchExpressions are ANDed? (As written, it's clear that each of them is internally ANDed, but not that the final result is obtained by ANDing the other two results.)

File: docs/user-guide/labels.md (hunk 9 above)
Comment (bgrant0607) [inline docs/user-guide/labels.md]: Exists isn't a keyword.

File: docs/user-guide/labels.md (hunk 4 above)
Comment (bgrant0607) [inline docs/user-guide/labels.md]: Reworded.

**Comment by bgrant0607** [comment]
Addressed comments and squashed, since it seems to be converging.

**Comment by bgrant0607** [comment]
@k8s-bot test this

**Comment by bgrant0607** [comment]
Shippable timeout was a flake.

**Comment by davidopp** [comment]
LGTM

**Pull request merged**